interface/src/payload.rs

```diff
@@ -0,0 +1,82 @@
+pub use parity_scale_codec::{Decode, Encode};
+
+use crate::MethodSelector;
+
+/// Execution payload is passed as input to VM to execute a metod.
+/// The optional state contains a serialized wallet state. It should be
+/// None if the call doesn't require any state.
+#[derive(Clone, Debug, Decode, Default, Encode, PartialEq, Eq)]
+pub struct ExecutionPayload {
+  pub state: Vec<u8>,
+  pub payload: Vec<u8>,
```

Comment:
I wasn't sure about this TBH. There is a tradeoff here: from the VM perspective it makes sense to put `Payload` type here, it simplifies the VM runner code. On the other hand, the host side always gets `payload` already **encoded** (encoded payload is part of a transaction). So, if `payload` was of type `Payload` instead of a `Vec<u8>`, the host side would have two options:

1. manually scale-encode Execution payload by encoding the raw state and appending the already encoded `payload` it got from TX. (A scale encoding of a struct is simply a concatenation of encodings of all its fields)
2. decode payload, put it into an `ExecutionPayload` struct, and encode the whole thing again.

In both cases the result is equivalent.

I optimized the code for host convenience here, but now that I think of it, it's probably wrong because it led to this weird `ExecutionPayload` with an opaque payload. I will change it to:
```rust
pub struct ExecutionPayload {
  pub state: Vec<u8>,
  pub payload: Payload,
}
```

and update the host code. Let's see which version is better.

Reply:
I think the change is fine. Also - we don't have to use the same exact data structures on both host and VM side!